src/core/utils/hooks/useCountdownTimer.ts

```diff
@@ -0,0 +1,51 @@
+import {useEffect, useRef, useState, useLayoutEffect} from "react";
+
+import {MINUTE_IN_MS} from "../time/timeConstants";
+import {CountdownResults} from "../time/timeTypes";
+import {calculateRemainingTime} from "../time/timeUtils";
+
+/**
+ * A React Hook that provides a countdown time
+ * @param {Date} target the target date we're counting down to
+ * @param {number} cadence the rate of the timer in milliseconds
+ *
+ * @returns {object} the CountDownResults object with ref
+ */
+function useCountDownTimer(target: Date, cadence = MINUTE_IN_MS): CountdownResults {
+  const interval = useRef<NodeJS.Timeout>();
+  const [countdown, setCountDown] = useState<CountdownResults>(
+    calculateRemainingTime(target)
+  );
+
+  useLayoutEffect(() => {
+    interval.current = setInterval(() => {
+      const data = calculateRemainingTime(target);
+
+      if (data.delta >= 0) {
+        setCountDown(data);
+      } else {
+        setCountDown({
+          delta: 0,
+          days: 0,
+          hours: 0,
+          minutes: 0,
+          seconds: 0
+        });
+      }
```

Comment:
I don't know why you need to add this else block. Interval seems to be cleared at the other effect below when  `data.delta <= 0`, does it not ?

Reply:
Yes, looks we don't need. I thought we can guarantee the trigger `onEnd` function but `clearInterval` works well instead.

Reply:
But I'm looking now, `clearInterval` is not working expectly. The following line doesn't execute and `countdown.delta` doesn't be zero. 
```
  if (data.delta >= 0) {
    setCountDown(data);
  }
```
So, in the `Countdown` component, the following line doesn't execute as well because `countdownData.delta` never is zero or negative value.
```
 useLayoutEffect(() => {
    if (countdownData.delta <= 0 && savedOnEndCallback.current) {
      savedOnEndCallback.current();
    }
  }, [countdownData.delta, onEnd]);
```
I guess, we should think something else.

Reply:
Weird but okay :+1: